src/Altinn.Platform/Altinn.Platform.Authentication/Authentication/Controllers/AuthenticationController.cs

```diff
@@ -340,8 +343,61 @@ private async Task<ActionResult> AuthenticateMaskinportenToken(string originalTo
                     }
                 }
 
+                string authenticatemethod = "maskinporten";
+
+                if (!string.IsNullOrEmpty(Request.Headers["X-Altinn-EnterpriseUser-Authentication"]))
+                {
+                    string enterpriseUserHeader = Request.Headers["X-Altinn-EnterpriseUser-Authentication"];
+                    EnterpriseUserCredentials credentials;
+
+                    try
+                    {
+                        credentials = DecodeEnterpriseUserHeader(enterpriseUserHeader, orgNumber);
+                    }
+                    catch (Exception)
+                    {
+                        return StatusCode(400);
+                    }
+
+                    HttpResponseMessage response = await _enterpriseUserAuthenticationService.AuthenticateEnterpriseUser(credentials);
```

Comment:
this response handling seems a bit mistplaced. Have you considered returning a different type of object from the service? Could be an object that holds the userAuthenticationResult in the case of the 200 response, and the http response in the case of an unexpected result, or only the status code? A private method could be called in all cases where  userAuthenticationResult isn't populated withouth including that logic in this function.

In the app template we throw a platformException that contains the response, and we catch this whenever we call the service. See example of use [here ](https://github.com/Altinn/altinn-studio/blob/master/src/Altinn.Apps/AppTemplates/AspNet/Altinn.App.PlatformServices/Implementation/DataAppSI.cs) and [here ](https://github.com/Altinn/altinn-studio/blob/master/src/Altinn.Apps/AppTemplates/AspNet/Altinn.App.Api/Controllers/DataController.cs#L173)

Reply:
We need to be able to return different ActionResults based on what Bridge returns. Instead of wrapping the object holding the enterpriser and the actionresult for the error in another object, I suggest we extract this entire if-block except for the happy path to a helper that returns a tuple with either the error set as an ActionResult or the authenticated enterpriser user. The block dealing with adding claims can still be done in this method. Something like

```
private async Task<(UserAuthenticationResult, ActionResult)> HandleEnterpriseUserLogin()
{
    UserAuthenticationResult authenticatedEnterpriseUser = null;
    ActionResult error = null;

    if (!string.IsNullOrEmpty(Request.Headers["X-Altinn-EnterpriseUser-Authentication"]))
    {
        // call, helper, service. Populate either authenticatedEnterpriseUser or error
    }

    return (authenticatedEnterpriseUser, error);
}
...

//  within AuthenticateMaskinportenToken

(UserAuthenticationResult authenticatedEnterpriseUser, ActionResult error) = await HandleEnterpriseUserLogin();

if (error != null)
{
    // Some error occured, f.ex. 429 Too Many Requests with a Retry-At-header
    return error;
}

if (authenticatedEnterpriseUser != null)
{
    // successful login, add claims
}

```

Reply:
Implemented the solution in example by calling HandleEnterpriseUserLogin and letting it handle the responses from SblBridge.